cblox_ros/include/cblox_ros/submap_server_inl.h

```diff
@@ -0,0 +1,802 @@
+}
+
+template<typename SubmapType>
+bool SubmapServer<SubmapType>::loadMap(const std::string& file_path) {
+  bool success = io::LoadSubmapCollection<SubmapType>(
+      file_path, &submap_collection_ptr_);
+  if (success) {
+    ROS_INFO("[CbloxServer] Successfully loaded SubmapCollection.");
+    constexpr bool kVisualizeMapOnLoad = true;
+    if (kVisualizeMapOnLoad) {
+      ROS_INFO("[CbloxServer] Publishing loaded map's mesh.");
+      visualizeWholeMap();
+    }
+  }
+  return success;
+}
+
+template<typename SubmapType>
+bool SubmapServer<SubmapType>::saveMapCallback(
+    voxblox_msgs::FilePath::Request& request,
+    voxblox_msgs::FilePath::Response& /*response*/) {
+  return saveMap(request.file_path);
+}
+
+template<typename SubmapType>
+bool SubmapServer<SubmapType>::loadMapCallback(
+    voxblox_msgs::FilePath::Request& request,
+    voxblox_msgs::FilePath::Response& /*response*/) {
+  bool success = loadMap(request.file_path);
+  return success;
+}
+
+template<typename SubmapType>
+void SubmapServer<SubmapType>::publishSubmap(
+    SubmapID submap_id) const {
+  if (submap_pub_.getNumSubscribers() == 0) {
+    return;
+  }
+  if (verbose_) {
+    ROS_INFO("[CbloxServer] Publishing submap %d", submap_id);
+  }
+
+  CHECK(submap_collection_ptr_->exists(submap_id))
+      << "[CbloxServer] Submap " << submap_id << " does not exist!";
+  cblox_msgs::MapLayer submap_msg;
+  serializeSubmapToMsg<SubmapType>(
+      submap_collection_ptr_->getSubmapPtr(submap_id), &submap_msg);
+  submap_pub_.publish(submap_msg);
+}
+
+template <typename SubmapType>
+bool SubmapServer<SubmapType>::publishActiveSubmapCallback(
+    cblox_msgs::SubmapSrvRequest& /*request*/,
+    cblox_msgs::SubmapSrvResponse& response) {
+  std::lock_guard<std::mutex> lock(visualizer_mutex_);
+
+  SubmapID submap_id = submap_collection_ptr_->getActiveSubmapID();
+  if (!submap_collection_ptr_->exists(submap_id)) {
+    ROS_ERROR("[CbloxServer] Active submap does not exist");
+    return false;
+  }
+  TsdfSubmap::Ptr submap_ptr =
+      submap_collection_ptr_->getSubmapPtr(submap_id);
+  if (submap_ptr->getTsdfMapPtr()->getTsdfLayerPtr()
+          ->getNumberOfAllocatedBlocks() == 0) {
+    if (verbose_) {
+      ROS_WARN("[CbloxServer] Active submap has no allocated blocks yet");
+    }
+    return false;
+  }
+
+  cblox_msgs::MapLayer submap_msg;
+  serializeSubmapToMsg<TsdfSubmap>(submap_ptr, &submap_msg);
+  response.submap_msg = submap_msg;
+  return true;
+}
+
+template<typename SubmapType>
+bool SubmapServer<SubmapType>::publishActiveSubmap() {
+  if (submap_collection_ptr_->empty()) {
+    ROS_WARN("[CbloxPlanner] Active submap does not exist!");
+    return false;
+  }
+  SubmapID submap_id = submap_collection_ptr_->getActiveSubmapID();
+  submap_collection_ptr_->getSubmapPtr(submap_id)->generateEsdf();
+  publishSubmap(submap_id);
+  return true;
+}
+template<>
+bool SubmapServer<TsdfSubmap>::publishActiveSubmap();
```

Comment:
I guess we have two implementations here because of the call to `generateEsdf()`?

because the two specializations share quite a lot of code, I would move keep this function unspecified and make another function, maybe `prepareSubmapForPublish(SubmapType submap)` and put the specialized code in there.

Reply:
I also made it virtual same as `finishSubmap`, although it is doing the same (generate esdf), I created a new method `prepareForPublish` since it might differ (from `finishSubmap`) in the derived class